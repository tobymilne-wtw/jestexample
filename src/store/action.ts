import { createAction, props } from '@ngrx/store';
import { Todo } from './todo.model';

const loadTodos = createAction('[Todo] Load Todos');
const loadTodosSuccess = createAction('[Todo] Load Todos Success', props<{ todos: Todo[]; }>());
const loadTodosFailure = createAction('[Todo] Load Todos Failure', props<{ error: string; }>());
const addTodo = createAction('[Todo] Add Todo', props<{ todo: Todo; }>());
const updateTodo = createAction('[Todo] Update Todo', props<{ todo: Todo; }>());
const deleteTodo = createAction('[Todo] Delete Todo', props<{ id: string; }>());

export const TodoActions = {    
    loadTodos,
    loadTodosSuccess,
    loadTodosFailure,
    addTodo,
    updateTodo,
    deleteTodo
};
