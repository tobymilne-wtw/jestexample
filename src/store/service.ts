import { Injectable } from '@angular/core';
import { Observable, delay, of } from 'rxjs';
import { Todo } from './todo.model';

@Injectable()
export class ToDoService {
 // fake backend
  getAll(): Observable<Todo[]> {
    return of(
      [{
        id: 1,
        description: 'description 1',
        completed: false
      },
      {
        id: 2,
        description: 'description 2',
        completed: false
      }]
    ).pipe(delay(2000))
  }
}
