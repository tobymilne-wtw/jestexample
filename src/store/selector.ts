import {  createSelector } from "@ngrx/store"
import { TodoState } from "./reducer";
import { AppState } from "./store";
 

const feature = (state: AppState) => state.todo;

export const todoSelector = createSelector(
  feature,
  (state: TodoState) => state.todos
);