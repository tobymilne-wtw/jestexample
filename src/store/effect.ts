import { Injectable } from '@angular/core';
import { Actions, createEffect, ofType } from '@ngrx/effects';
import { TodoActions } from './action';
import { catchError, map, mergeMap, of } from 'rxjs';
import { ToDoService } from './service';

@Injectable()
export class TodoEffects {

  loadTodos$ = createEffect(() =>
    this.actions$.pipe(
      ofType(TodoActions.loadTodos),
      mergeMap(() =>
        this.todoService.getAll().pipe(
          map((todos) => TodoActions.loadTodosSuccess({ todos })),
          catchError((error) =>
            of(TodoActions.loadTodosFailure({ error: error.message }))
          )
        )
      )
    )
  );
  constructor(private actions$: Actions, private todoService: ToDoService) {}
}
