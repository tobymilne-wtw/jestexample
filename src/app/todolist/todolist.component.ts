import { CommonModule } from '@angular/common';
import { ChangeDetectionStrategy, Component, Signal, OnInit } from '@angular/core';
import { Todo } from '../../store/todo.model';
import { AppState } from '../../store/store';
import { Store } from '@ngrx/store';
import { TodoActions } from '../../store/action';
import { todoSelector } from '../../store/selector';
import { toSignal } from '@angular/core/rxjs-interop'

@Component({
  selector: 'app-todolist',
  standalone: true,
  imports: [CommonModule],
  templateUrl: './todolist.component.html',
  styleUrl: './todolist.component.css',
  changeDetection: ChangeDetectionStrategy.OnPush,
})
export class TodolistComponent implements OnInit {
  todos$: Signal<Todo[] | undefined>;
  isLoading$: Signal<boolean | undefined>;

  constructor(public store: Store<AppState>) {
    this.todos$ = toSignal(this.store.select(todoSelector)); // observable converted to signal
    this.isLoading$ = toSignal(this.store.select((state) => state.todo.loading));
  }

  ngOnInit() {
   this.loadTodos();
  }

  loadTodos() {
    this.store.dispatch(TodoActions.loadTodos());
  }

  addTodo(index: number) {
    const todo: Todo = { id: index, description: `description ${index}`, completed: false };
    this.store.dispatch(TodoActions.addTodo({ todo }));
  }

  complete(todo: Todo) {
    this.store.dispatch(
      TodoActions.updateTodo({ todo: { ...todo, completed: true } })
    );
  }
 }


