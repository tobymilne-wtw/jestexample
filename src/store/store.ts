import { Action, ActionReducer } from '@ngrx/store';
import { TodoState, todoReducer } from './reducer';
import { TodoEffects } from './effect';

export interface AppState {
    todo: TodoState
  }
  
  export interface AppStore {
    todo: ActionReducer<TodoState, Action>;
  }
  
  export const appStore: AppStore = {
    todo: todoReducer
  }
  
  export const appEffects = [TodoEffects];
  