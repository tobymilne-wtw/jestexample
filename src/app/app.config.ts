import { ApplicationConfig } from '@angular/core';
import { provideRouter } from '@angular/router';

import { routes } from './app.routes';
import { provideStore } from '@ngrx/store';
import { provideEffects } from '@ngrx/effects';
import { appEffects, appStore } from '../store/store';
import { ToDoService } from '../store/service';
import { TodolistComponent } from './todolist/todolist.component';

export const appConfig: ApplicationConfig = {
  providers: [provideRouter(routes),
  provideStore(appStore),
    provideEffects(appEffects),
    ToDoService,
  TodolistComponent]
};
