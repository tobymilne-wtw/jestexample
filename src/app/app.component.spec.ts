import { TestBed } from '@angular/core/testing';
import { AppComponent } from './app.component';
import { provideMockStore } from '@ngrx/store/testing';
import { TodolistComponent } from './todolist/todolist.component';
import { AppState } from '../store/store';


const initialState: AppState = {todo: {todos:[], loading: false, error: ''}};

describe('AppComponent', () => {
  beforeEach(async () => {
    await TestBed.configureTestingModule({
      imports: [AppComponent, TodolistComponent],
      providers: [provideMockStore({initialState})]
    }).compileComponents();
  });

  test('create the app', () => {
    const fixture = TestBed.createComponent(AppComponent);
    const app = fixture.componentInstance;
    expect(app).toBeTruthy();
  });

  test(`have the 'testsite' title`, () => {
    const fixture = TestBed.createComponent(AppComponent);
    const app = fixture.componentInstance;
    expect(app.title).toEqual('testsite');
  });

  test.skip('render title', () => {
    const fixture = TestBed.createComponent(AppComponent);
    fixture.detectChanges();
    const compiled = fixture.nativeElement as HTMLElement;
    expect(compiled.querySelector('h1')?.textContent).toContain('Hello, testsite');
  });

  test('render correctly', () => {
    const fixture = TestBed.createComponent(AppComponent);
    fixture.detectChanges();
   
    expect(fixture).toMatchSnapshot();
  });
});
