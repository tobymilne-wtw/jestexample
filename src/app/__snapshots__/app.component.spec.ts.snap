// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppComponent render correctly 1`] = `
<app-root
  title={[Function String]}
>
  <main
    class="main"
  >
    <app-todolist>
      <div>
        No Data
      </div>
      <button>
        Refresh
      </button>
      <button>
        Add
      </button>
    </app-todolist>
  </main><router-outlet />
</app-root>
`;

exports[`AppComponent should render correctly 1`] = `
<app-root
  title={[Function String]}
>
  <main
    class="main"
  >
    <app-todolist>
      <div>
        No Data
      </div>
      <button>
        Refresh
      </button>
      <button>
        Add
      </button>
    </app-todolist>
  </main><router-outlet />
</app-root>
`;
